### PR TITLE
Parametrize dragon install

### PR DESCRIPTION
**Pull request by ankona**
Parameterize the `smart build --dragon` command to enable specification of a fork/repository and package version

- add parameter `--dragon-repo`
- add parameter `--dragon-version`

Sample usage:

- Get latest version from private fork

  `GH_TOKEN=xxxxx smart build --dragon-repo ankona/dragonfork`

- Get specific version from private fork

  `GH_TOKEN=xxxxx smart build --dragon-repo ankona/dragonfork --dragon-version 0.10`

- Get specific version from public fork

  - `GH_TOKEN=xxxxx smart build --dragon-repo dragonhpc/dragon --dragon-version 0.10`
  - `GH_TOKEN=xxxxx smart build --dragon-version 0.10`
  - `smart build --dragon-repo dragonhpc/dragon --dragon-version 0.10`
  - `smart build --dragon-version 0.10`

## manual test results

1. OK - `smart build --dragon`

```
  [1] % smart build --dragon
  [SmartSim] INFO Running SmartSim build process...
  [SmartSim] INFO Checking requested versions...
  [SmartSim] DEBUG Checking for build tools...
  [SmartSim] DEBUG Retrieved asset metadata: GitReleaseAsset(url="https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358")
  [SmartSim] DEBUG Retrieved asset dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz metadata from https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358
  [SmartSim] DEBUG Extracted dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz to /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] DEBUG Installing packages:
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/dragon-0.9-cp311-cp311-linux_x86_64.whl
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/pycapnp-2.0.0-cp311-cp311-linux_x86_64.whl
  [SmartSim] DEBUG Deleted temporary files in: /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] INFO Dragon installation complete
  [SmartSim] INFO Redis build complete!
```

2. OK_FAIL - `smart build --dragon-repo dragonhpc/dragon-nightly`

```
  [1] % smart build --dragon-repo dragonhpc/dragon-nightly
  [SmartSim] INFO Running SmartSim build process...
  [SmartSim] INFO Checking requested versions...
  [SmartSim] DEBUG Checking for build tools...
  [SmartSim] ERROR An access token must be available to access dragonhpc/dragon-nightly. Set the `GH_TOKEN` env var to pass your access token.
  [SmartSim] WARNING Dragon installation failed
  [SmartSim] INFO Redis build complete!
```

3. OK_FAIL - `smart build --dragon-repo dragonhpc/dragon-nightly --dragon-version 0.9`
  - no gh_token supplied to connect to private repo, no version 0.9 in that repo

```
  [130] % smart build --dragon-repo dragonhpc/dragon-nightly --dragon-version 0.9 
  [SmartSim] INFO Running SmartSim build process...
  [SmartSim] INFO Checking requested versions...
  [SmartSim] DEBUG Checking for build tools...
  [SmartSim] ERROR An access token must be available to access dragonhpc/dragon-nightly. Set the `GH_TOKEN` env var to pass your access token.
  [SmartSim] WARNING Dragon installation failed
  [SmartSim] INFO Redis build complete!
```

3. OK_FAIL - `GH_TOKEN=xxx smart build --dragon-repo dragonhpc/dragon-nightly --dragon-version 0.9`
  - no version 0.9 in that repo

```
  [SmartSim] INFO Running SmartSim build process...
  [SmartSim] INFO Checking requested versions...
  [SmartSim] DEBUG Checking for build tools...
  [SmartSim] WARNING Please specify a dragon version (e.g. 0.9) of an asset available in the repository:
          dragon-0.10-py3.10.10-ff4c77a60.tar.gz
          dragon-0.10-py3.11.5-ff4c77a60.tar.gz
          dragon-0.10-py3.9.4.1-ff4c77a60.tar.gz
          dragondocs-0.10-ff4c77a60.tar.gz
  [SmartSim] WARNING No dragon runtime asset available to install
  [SmartSim] WARNING Dragon installation failed
  [SmartSim] INFO Redis build complete!
```

4. OK FAIL - `smart build --dragon-repo dragonhpc/dragon-nightly --dragon-version 0.10`
  - no gh_token supplied to connect to private repo

```
  [1] % smart build --dragon-repo dragonhpc/dragon-nightly --dragon-version 0.10
  [SmartSim] INFO Running SmartSim build process...
  [SmartSim] INFO Checking requested versions...
  [SmartSim] DEBUG Checking for build tools...
  [SmartSim] ERROR An access token must be available to access dragonhpc/dragon-nightly. Set the `GH_TOKEN` env var to pass your access token.
  [SmartSim] WARNING Dragon installation failed
  [SmartSim] INFO Redis build complete!
```

5. OK - `GH_TOKEN=xxx  smart build --dragon-repo dragonhpc/dragon-nightly --dragon-version 0.10`

```
  [SmartSim] INFO Running SmartSim build process...
  [SmartSim] INFO Checking requested versions...
  [SmartSim] DEBUG Checking for build tools...
  [SmartSim] WARNING Platform-specific package not found. Using dragon-0.10-py3.11.5-ff4c77a60.tar.gz
  [SmartSim] DEBUG Retrieved asset metadata: GitReleaseAsset(url="https://api.github.com/repos/DragonHPC/dragon-nightly/releases/assets/190659388")
  [SmartSim] DEBUG Retrieved asset dragon-0.10-py3.11.5-ff4c77a60.tar.gz metadata from https://api.github.com/repos/DragonHPC/dragon-nightly/releases/assets/190659388
  [SmartSim] DEBUG Extracted dragon-0.10-py3.11.5-ff4c77a60.tar.gz to /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/190659388
  [SmartSim] DEBUG Installing packages:
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/190659388/dragon-0.10-py3.11.5-ff4c77a60.tar.gz/dragon-0.10/pycapnp-2.0.0-cp311-cp311-linux_x86_64.whl
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/190659388/dragon-0.10-py3.11.5-ff4c77a60.tar.gz/dragon-0.10/dragon-0.10-cp311-cp311-linux_x86_64.whl
  [SmartSim] DEBUG Deleted temporary files in: /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/190659388
  [SmartSim] INFO Dragon installation complete
  [SmartSim] INFO Redis build complete!
```

5. OK_FAIL - `smart build --dragon-version 0.10`
  - no v0.10 in that repo

```
  [SmartSim] INFO Running SmartSim build process...
  [SmartSim] INFO Checking requested versions...
  [SmartSim] DEBUG Checking for build tools...
  [SmartSim] WARNING Please specify a dragon version (e.g. 0.9) of an asset available in the repository:
          dragon-0.9-py3.10.10-CRAYEX-ce895d2de.tar.gz
          dragon-0.9-py3.10.10-ec3fc0f8a.tar.gz
          dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz
          dragon-0.9-py3.11.5-ec3fc0f8a.tar.gz
          dragon-0.9-py3.9.4.1-CRAYEX-ce895d2de.tar.gz
          dragon-0.9-py3.9.4.1-ec3fc0f8a.tar.gz
  [SmartSim] WARNING No dragon runtime asset available to install
  [SmartSim] WARNING Dragon installation failed
  [SmartSim] INFO Redis build complete!
```

6. OK - `smart build --dragon-version 0.9`

```
  [SmartSim] INFO Running SmartSim build process...
  [SmartSim] INFO Checking requested versions...
  [SmartSim] DEBUG Checking for build tools...
  [SmartSim] DEBUG Retrieved asset metadata: GitReleaseAsset(url="https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358")
  [SmartSim] DEBUG Retrieved asset dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz metadata from https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358
  [SmartSim] DEBUG Extracted dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz to /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] DEBUG Installing packages:
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/dragon-0.9-cp311-cp311-linux_x86_64.whl
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/pycapnp-2.0.0-cp311-cp311-linux_x86_64.whl
  [SmartSim] DEBUG Deleted temporary files in: /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] INFO Dragon installation complete
  [SmartSim] INFO Redis build complete!
```

7. OK - `export GH_TOKEN=xxx smart build --dragon-version 0.9`

```
  [SmartSim] INFO Running SmartSim build process...
  [SmartSim] INFO Checking requested versions...
  [SmartSim] DEBUG Checking for build tools...
  [SmartSim] DEBUG Retrieved asset metadata: GitReleaseAsset(url="https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358")
  [SmartSim] DEBUG Retrieved asset dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz metadata from https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358
  [SmartSim] DEBUG Extracted dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz to /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] DEBUG Installing packages:
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/dragon-0.9-cp311-cp311-linux_x86_64.whl
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/pycapnp-2.0.0-cp311-cp311-linux_x86_64.whl
  [SmartSim] DEBUG Deleted temporary files in: /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] INFO Dragon installation complete
  [SmartSim] INFO Redis build complete!
```

8. OK - `smart build --dragon-repo dragonhpc/dragon`
  - handle lower case fine!

```
  [1] % smart build --dragon-repo dragonhpc/dragon
  [SmartSim] INFO Running SmartSim build process...
  [SmartSim] INFO Checking requested versions...
  [SmartSim] DEBUG Checking for build tools...
  [SmartSim] DEBUG Retrieved asset metadata: GitReleaseAsset(url="https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358")
  [SmartSim] DEBUG Retrieved asset dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz metadata from https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358
  [SmartSim] DEBUG Extracted dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz to /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] DEBUG Installing packages:
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/dragon-0.9-cp311-cp311-linux_x86_64.whl
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/pycapnp-2.0.0-cp311-cp311-linux_x86_64.whl
  [SmartSim] DEBUG Deleted temporary files in: /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] INFO Dragon installation complete
  [SmartSim] INFO Redis build complete!
```

9. OK - `GH_TOKEN=xxx smart build --dragon-repo dragonhpc/dragon`
  - token not required for public, but works:

```
  [SmartSim] INFO Running SmartSim build process...
  [SmartSim] INFO Checking requested versions...
  [SmartSim] DEBUG Checking for build tools...
  [SmartSim] DEBUG Retrieved asset metadata: GitReleaseAsset(url="https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358")
  [SmartSim] DEBUG Retrieved asset dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz metadata from https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358
  [SmartSim] DEBUG Extracted dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz to /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] DEBUG Installing packages:
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/dragon-0.9-cp311-cp311-linux_x86_64.whl
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/pycapnp-2.0.0-cp311-cp311-linux_x86_64.whl
  [SmartSim] DEBUG Deleted temporary files in: /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] INFO Dragon installation complete
  [SmartSim] INFO Redis build complete!
```

8. OK - `smart build --dragon-repo dragonhpc/dragon --dragon-version 0.9`

```
  [1] % smart build --dragon-repo dragonhpc/dragon --dragon-version 0.9 
  [SmartSim] INFO Running SmartSim build process...
  [SmartSim] INFO Checking requested versions...
  [SmartSim] DEBUG Checking for build tools...
  [SmartSim] DEBUG Retrieved asset metadata: GitReleaseAsset(url="https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358")
  [SmartSim] DEBUG Retrieved asset dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz metadata from https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358
  [SmartSim] DEBUG Extracted dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz to /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] DEBUG Installing packages:
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/dragon-0.9-cp311-cp311-linux_x86_64.whl
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/pycapnp-2.0.0-cp311-cp311-linux_x86_64.whl
  [SmartSim] DEBUG Deleted temporary files in: /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] INFO Dragon installation complete
  [SmartSim] INFO Redis build complete!
```

9. OK - `GH_TOKEN=xxx  smart build --dragon-repo dragonhpc/dragon --dragon-version 0.9`

```
  [SmartSim] INFO Running SmartSim build process...
  [SmartSim] INFO Checking requested versions...
  [SmartSim] DEBUG Checking for build tools...
  [SmartSim] DEBUG Retrieved asset metadata: GitReleaseAsset(url="https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358")
  [SmartSim] DEBUG Retrieved asset dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz metadata from https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358
  [SmartSim] DEBUG Extracted dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz to /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] DEBUG Installing packages:
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/dragon-0.9-cp311-cp311-linux_x86_64.whl
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/pycapnp-2.0.0-cp311-cp311-linux_x86_64.whl
  [SmartSim] DEBUG Deleted temporary files in: /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] INFO Dragon installation complete
  [SmartSim] INFO Redis build complete!
```

10. OK - 1. `smart build --dragon --dragon-version 0.9`

```
  [SmartSim] INFO Running SmartSim build process...
  [SmartSim] INFO Checking requested versions...
  [SmartSim] DEBUG Checking for build tools...
  [SmartSim] DEBUG Retrieved asset metadata: GitReleaseAsset(url="https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358")
  [SmartSim] DEBUG Retrieved asset dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz metadata from https://api.github.com/repos/DragonHPC/dragon/releases/assets/165524358
  [SmartSim] DEBUG Extracted dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz to /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] DEBUG Installing packages:
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/dragon-0.9-cp311-cp311-linux_x86_64.whl
          /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358/dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz/dragon-0.9/pycapnp-2.0.0-cp311-cp311-linux_x86_64.whl
  [SmartSim] DEBUG Deleted temporary files in: /lus/bnchlu1/mcbridch/code/ss/smartsim/_core/.dragon/165524358
  [SmartSim] INFO Dragon installation complete
  [SmartSim] INFO Redis build complete!
```

11. OK_FAIL - 1. `smart build --dragon --dragon-version 0.10`

```
  [1] % smart build --dragon --dragon-version 0.10
  [SmartSim] INFO Running SmartSim build process...
  [SmartSim] INFO Checking requested versions...
  [SmartSim] DEBUG Checking for build tools...
  [SmartSim] WARNING Please specify a dragon version (e.g. 0.9) of an asset available in the repository:
          dragon-0.9-py3.10.10-CRAYEX-ce895d2de.tar.gz
          dragon-0.9-py3.10.10-ec3fc0f8a.tar.gz
          dragon-0.9-py3.11.5-CRAYEX-ce895d2de.tar.gz
          dragon-0.9-py3.11.5-ec3fc0f8a.tar.gz
          dragon-0.9-py3.9.4.1-CRAYEX-ce895d2de.tar.gz
          dragon-0.9-py3.9.4.1-ec3fc0f8a.tar.gz
  [SmartSim] WARNING No dragon runtime asset available to install
  [SmartSim] WARNING Dragon installation failed
  [SmartSim] INFO Redis build complete!

```

12. OK - `smart build -h`

```
  --dragon-repo DRAGON_REPO
                        Specify a git repo containing dragon release assets (e.g. DragonHPC/dragon)
  --dragon-version DRAGON_VERSION
                        Specify the dragon version to install (e.g. 0.9)
```

### CODE REVIEW

File: smartsim/_core/_cli/scripts/dragon_install.py [hunk 1]
```diff
@@ -77,39 +145,70 @@ def _version_filter(asset_name: str) -> bool:
...
+    all_releases = [release for release in list(dragon_repo.get_releases())]
+    return all_releases
 
-def _get_release_assets() -> t.Collection[GitReleaseAsset]:
+
+def _get_release_assets(request: DragonInstallRequest) -> t.Collection[GitReleaseAsset]:
     """Retrieve a collection of available assets for all releases that satisfy
     the dragon version pin
 
+    :param git_repo: The name of the git repository to search for assets, e.g. "DragonHPC/dragon"
```
Comment (AlyssaCote) [inline smartsim/_core/_cli/scripts/dragon_install.py:168]: `:param request:`

File: smartsim/_core/_cli/scripts/dragon_install.py [hunk 2]
```diff
@@ -77,39 +145,70 @@ def _version_filter(asset_name: str) -> bool:
...
     :param asset_name: A value to inspect for keywords indicating a dragon version
     :returns: True if supplied value is correct for current dragon version"""
-    return f"dragon-{dragon_pin()}" in asset_name
+    return f"dragon-{dragon_version}" in asset_name
+
+
+def _get_all_releases(dragon_repo: Repository) -> t.Collection[GitRelease]:
+    """Retrieve all available releases for the configured dragon repository
 
+    :returns: A list of GitRelease"""
```
Comment (AlyssaCote) [inline smartsim/_core/_cli/scripts/dragon_install.py:159]: `:param dragon_repo:` docstring

File: smartsim/_core/_cli/scripts/dragon_install.py [hunk 3]
```diff
@@ -77,39 +145,70 @@ def _version_filter(asset_name: str) -> bool:
...
 
 
-def _pin_filter(asset_name: str) -> bool:
+def _pin_filter(asset_name: str, dragon_version: str) -> bool:
     """Return true if the supplied value contains a dragon version pin match
 
     :param asset_name: A value to inspect for keywords indicating a dragon version
     :returns: True if supplied value is correct for current dragon version"""
-    return f"dragon-{dragon_pin()}" in asset_name
+    return f"dragon-{dragon_version}" in asset_name
```
Comment (AlyssaCote) [inline smartsim/_core/_cli/scripts/dragon_install.py:153]: `:param dragon_version:` docstring

File: smartsim/_core/_cli/scripts/dragon_install.py [hunk 4]
```diff
@@ -118,26 +217,42 @@ def filter_assets(assets: t.Collection[GitReleaseAsset]) -> t.Optional[GitReleas
...
+
     return asset
 
 
-def retrieve_asset_info() -> GitReleaseAsset:
+def retrieve_asset_info(request: DragonInstallRequest) -> GitReleaseAsset:
     """Find a release asset that meets all necessary filtering criteria
 
-    :param dragon_pin: identify the dragon version to install (e.g. dragon-0.8)
+    :param git_repo: The name of the git repository to search for assets, e.g. "DragonHPC/dragon"
```
Comment (AlyssaCote) [inline smartsim/_core/_cli/scripts/dragon_install.py:252]: `:param request:`

File: smartsim/_core/_cli/scripts/dragon_install.py [hunk 5]
```diff
@@ -152,40 +267,56 @@ def retrieve_asset_info() -> GitReleaseAsset:
     return asset
 
 
-def retrieve_asset(working_dir: pathlib.Path, asset: GitReleaseAsset) -> pathlib.Path:
+def retrieve_asset(
+    request: DragonInstallRequest, asset: GitReleaseAsset
+) -> pathlib.Path:
```
Comment (AlyssaCote) [inline smartsim/_core/_cli/scripts/dragon_install.py:272]: Not this line, but `:param request:`

File: smartsim/_core/_cli/scripts/dragon_install.py [hunk 6]
```diff
@@ -152,40 +267,56 @@ def retrieve_asset_info() -> GitReleaseAsset:
...
 
     # extract the asset
-    archive = WebTGZ(download_url)
-    archive.extract(download_dir)
+    try:
+        archive = _WebTGZ(presigned_url, headers=headers)
+        archive.extract(asset_path)
+        logger.debug(f"Extracted {asset.name} to {download_dir}")
+    except Exception as ex:
+        logger.exception("oops")
```
Comment (AlyssaCote) [inline smartsim/_core/_cli/scripts/dragon_install.py:314]: "oops" 😂

File: smartsim/_core/_cli/scripts/dragon_install.py (hunk 5 above)
Comment (AlyssaCote) [inline smartsim/_core/_cli/scripts/dragon_install.py:272]: Oh and can you also add a `:raises:` docstring?

File: smartsim/_core/_cli/scripts/dragon_install.py [hunk 7]
```diff
@@ -152,40 +267,56 @@ def retrieve_asset_info() -> GitReleaseAsset:
...
 
-    logger.debug(f"Extracted {download_url} to {download_dir}")
     return download_dir
 
 
-def install_package(asset_dir: pathlib.Path) -> int:
+def install_package(request: DragonInstallRequest, asset_dir: pathlib.Path) -> int:
```
Comment (AlyssaCote) [inline smartsim/_core/_cli/scripts/dragon_install.py:319]: another "fix up the docstring" request here as well :)

**Review by AlyssaCote — CHANGES_REQUESTED** [review]
This looks really good to me! Just a couple docstring fix requests and a question about to what extent we should document this, but once the tests pass I think this should be good to go.

File: smartsim/_core/_cli/scripts/dragon_install.py (hunk 1 above)
Comment (ankona) [reply smartsim/_core/_cli/scripts/dragon_install.py:168]: fixed 3x where copypasta got me

File: smartsim/_core/_cli/scripts/dragon_install.py (hunk 3 above)
Comment (ankona) [reply smartsim/_core/_cli/scripts/dragon_install.py:153]: fixed

File: smartsim/_core/_cli/scripts/dragon_install.py (hunk 4 above)
Comment (ankona) [reply smartsim/_core/_cli/scripts/dragon_install.py:252]: fixed

Comment (ankona) [reply smartsim/_core/_cli/scripts/dragon_install.py:252]: fixed!

File: smartsim/_core/_cli/scripts/dragon_install.py (hunk 5 above)
Comment (ankona) [reply smartsim/_core/_cli/scripts/dragon_install.py:272]: fixed!

File: smartsim/_core/_cli/scripts/dragon_install.py (hunk 6 above)
Comment (ankona) [reply smartsim/_core/_cli/scripts/dragon_install.py:314]: re-raised as smartsim error

File: smartsim/_core/_cli/scripts/dragon_install.py (hunk 5 above)
Comment (ankona) [reply smartsim/_core/_cli/scripts/dragon_install.py:272]: I can!

File: smartsim/_core/_cli/scripts/dragon_install.py (hunk 7 above)
Comment (ankona) [reply smartsim/_core/_cli/scripts/dragon_install.py:319]: got it!

**Comment by al-rigazzi** [comment]
@AlyssaCote do you think we need to add the `dragon-config` step to install?

**Comment by al-rigazzi** [comment]
@AlyssaCote yeah, not for tests, you're right. I guess, with the current implementation, if a user wanted to use HSTAs, the `dragon-config` step with appropriate parameters should be run from the installation directories, do you agree?

**Comment by AlyssaCote** [comment]
@al-rigazzi Ohhh good point. I was only thinking of the green checks for our github actions! Yeah you're right we do need a way for users to enable HSTA.

File: smartsim/_core/_cli/build.py [hunk 8]
```diff
@@ -483,6 +498,21 @@ def configure_parser(parser: argparse.ArgumentParser) -> None:
...
+            "Specify a git repo containing dragon release assets "
+            f"(e.g. {DEFAULT_DRAGON_REPO})"
+        ),
+    )
+    parser.add_argument(
+        "--dragon-version",
+        default=None,
+        type=str,
+        help=f"Specify the dragon version to install (e.g. {DEFAULT_DRAGON_VERSION})",
+    )
```
[the remark this answers is not in the document]
Comment (ankona) [reply smartsim/_core/_cli/build.py:515]: Added a blurb on the `dragon_install` docs page

File: smartsim/_core/_cli/scripts/dragon_install.py [hunk 9]
```diff
@@ -77,67 +145,117 @@ def _version_filter(asset_name: str) -> bool:
...
+
     return asset
 
 
-def retrieve_asset_info() -> GitReleaseAsset:
+def retrieve_asset_info(request: DragonInstallRequest) -> GitReleaseAsset:
     """Find a release asset that meets all necessary filtering criteria
 
-    :param dragon_pin: identify the dragon version to install (e.g. dragon-0.8)
+    :param request: details of a requesst for the installation of the dragon package
```
Comment (AlyssaCote) [inline smartsim/_core/_cli/scripts/dragon_install.py:255]: `requesst` -> `request`

Comment (AlyssaCote) [reply smartsim/_core/_cli/scripts/dragon_install.py:255]: it's in a couple places

Comment (ankona) [reply smartsim/_core/_cli/scripts/dragon_install.py:255]: I'm channeling @MattToast here!

File: smartsim/_core/_install/builder.py [hunk 10]
```diff
@@ -795,50 +795,92 @@ def __place_for_rai__(self, target: _PathLike) -> Path:
...
+        super().__init__(headers)
+        self._url = url
+
+    @property
+    def url(self) -> str:
+        """Returns the url that was downloaded"""
+        return self._url
+
     def _extract_download(self, download_path: Path, target: _PathLike) -> None:
+        """Called during file handling to perform extraction of `.tar.gz` files"""
```
Comment (AlyssaCote) [inline smartsim/_core/_install/builder.py:869]: docstrings

File: smartsim/_core/_install/builder.py [hunk 11]
```diff
@@ -795,50 +795,92 @@ def __place_for_rai__(self, target: _PathLike) -> Path:
...
         with tarfile.open(download_path, "r") as tgz_file:
             tgz_file.extractall(target)
 
 
 class _WebZip(_ExtractableWebArchive):
+    """Performs download and extraction of a remote archive file
+    in the `.zip` format."""
+
     def _extract_download(self, download_path: Path, target: _PathLike) -> None:
+        """Called during file handling to perform extraction of `.zip` files"""
```
Comment (AlyssaCote) [inline smartsim/_core/_install/builder.py:879]: docstrings

File: smartsim/_core/_install/builder.py [hunk 12]
```diff
@@ -795,50 +795,92 @@ def __place_for_rai__(self, target: _PathLike) -> Path:
...
+    extraction of a remote archive file"""
+
     @abstractmethod
-    def _extract_download(self, download_path: Path, target: _PathLike) -> None: ...
+    def _extract_download(self, download_path: Path, target: _PathLike) -> None:
+        """Called during file handling to perform format-specific extraction
+        operations. Must be overridden in child classes"""
 
     def extract(self, target: _PathLike) -> None:
+        """Extract the downloaded file into the desired target location"""
```
Comment (AlyssaCote) [inline smartsim/_core/_install/builder.py:845]: docstrings

**Review by AlyssaCote — APPROVED** [review]
LGTM! Can't wait to see all the green check marks!!

File: smartsim/_core/_install/builder.py (hunk 10 above)
Comment (ankona) [reply smartsim/_core/_install/builder.py:869]: fixed!

File: smartsim/_core/_install/builder.py (hunk 11 above)
Comment (ankona) [reply smartsim/_core/_install/builder.py:879]: fixed!

File: smartsim/_core/_install/builder.py (hunk 12 above)
Comment (ankona) [reply smartsim/_core/_install/builder.py:845]: fixed!

File: smartsim/_core/_cli/scripts/dragon_install.py (hunk 9 above)
Comment (MattToast) [reply smartsim/_core/_cli/scripts/dragon_install.py]: I the reader gets the gist, its good enough in my book! 😉

**Pull request merged**